mir-opt: Ignore the dead store statements in MatchBranchSimplification

dianqk opened the pull request:
#128299 simplified

```rust
bb0: {
    _2 = discriminant((*_1));
    switchInt(move _2) -> [0: bb4, 1: bb3, otherwise: bb2];
}
bb3: {
    _5 = &(((*_1) as B).0: NestCopy);
    _7 = copy ((((*_1) as B).0: NestCopy).0: i32);
    _9 = &((((*_1) as B).0: NestCopy).1: AllCopy);
    _11 = copy (((((*_1) as B).0: NestCopy).1: AllCopy).0: i32);
    _12 = copy (((((*_1) as B).0: NestCopy).1: AllCopy).1: u64);
    _13 = copy (((((*_1) as B).0: NestCopy).1: AllCopy).2: [i8; 3]);
    _8 = AllCopy { a: move _11, b: move _12, c: move _13 };
    _10 = copy ((((*_1) as B).0: NestCopy).2: [i8; 3]);
    _6 = NestCopy { a: move _7, b: move _8, c: move _10 };
    _0 = Enum1::B(move _6);
    goto -> bb1;
}
bb4: {
    _3 = &(((*_1) as A).0: AllCopy);
    _14 = copy ((((*_1) as A).0: AllCopy).0: i32);
    _15 = copy ((((*_1) as A).0: AllCopy).1: u64);
    _16 = copy ((((*_1) as A).0: AllCopy).2: [i8; 3]);
    _4 = AllCopy { a: move _14, b: move _15, c: move _16 };
    _0 = Enum1::A(move _4);
    StorageDead(_4);
    goto -> bb1;
}
```
to
```rust
bb0: {
    _2 = discriminant((*_1));
    switchInt(move _2) -> [0: bb1, 1: bb2, otherwise: bb4];
}
bb1: {
    _3 = &(((*_1) as A).0: AllCopy);
    _0 = copy (*_1);
    goto -> bb3;
}
bb2: {
    _4 = &(((*_1) as B).0: NestCopy);
    StorageLive(_5);
    _5 = &((((*_1) as B).0: NestCopy).1: AllCopy);
    StorageDead(_5);
    _0 = copy (*_1);
    goto -> bb3;
}
```

Since all of these are dead store statements, we can simplify and merge them into a single copy statement.

DianQK commented:
r? ghost

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10684205183/job/29614084174) [(plain)](https://github.com/rust-lang/rust/commit/efce2e86595b53291bbbe609ff4138d05509e745/checks/29614084174/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.904 Building wheels for collected packages: reuse
#13 2.905   Building wheel for reuse (pyproject.toml): started
#13 3.155   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.157   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.157   Stored in directory: /tmp/pip-ephem-wheel-cache-3id33myr/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.159 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.559 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.560 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.097 Collecting virtualenv
#13 4.097 Collecting virtualenv
#13 4.153   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.383      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 24.9 MB/s eta 0:00:00
#13 4.446 Collecting filelock<4,>=3.12.2
#13 4.454   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.488 Collecting platformdirs<5,>=3.9.1
#13 4.496   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.518 Collecting distlib<1,>=0.3.7
#13 4.537      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 49.3 MB/s eta 0:00:00
#13 4.537      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 49.3 MB/s eta 0:00:00
#13 4.618 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.807 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.9s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      184256 kB
DirectMap2M:     7155712 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/d6c8169c186ab16a3404cd0d0866674018e8a19e/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-d6c8169c186ab16a3404cd0d0866674018e8a19e-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 28.97s
##[endgroup]
fmt check
Diff in /checkout/tests/mir-opt/match_br/match_br_merge_copy.rs:70:
     // CHECK-NEXT return;
     match a {
         Some(_b) => a,
+        None => None,
     }
 }
 
 
Diff in /checkout/tests/mir-opt/match_br/match_br_merge_copy.rs:84:
         Some(_) => {
         Some(_) => {
             *b = 1;
-        },
-        None => None
+        }
+        None => None,
+        None => None,
     }
 }
 
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/mir-opt/building/match/match_false_edges.rs" "/checkout/tests/mir-opt/building/match/simple_match.rs" "/checkout/tests/mir-opt/building/match/exponential_or.rs" "/checkout/tests/mir-opt/building/issue_110508.rs" "/checkout/tests/mir-opt/building/eq_never_type.rs" "/checkout/tests/mir-opt/building/storage_live_dead_in_statics.rs" "/checkout/tests/mir-opt/building/receiver_ptr_mutability.rs" "/checkout/tests/mir-opt/building/logical_or_in_conditional.rs" "/checkout/tests/mir-opt/match_br/match_br_merge_copy.rs" "/checkout/tests/mir-opt/byte_slice.rs" "/checkout/tests/mir-opt/switch_to_self.rs" "/checkout/tests/mir-opt/issue_78192.rs" "/checkout/tests/mir-opt/coroutine_storage_dead_unwind.rs" "/checkout/tests/mir-opt/reference_prop.rs" "/checkout/tests/mir-opt/issue_38669.rs" "/checkout/tests/mir-opt/issue_101973.rs" "/checkout/tests/mir-opt/remove_storage_markers.rs" "/checkout/tests/mir-opt/inline_coroutine_body.rs" "/checkout/tests/mir-opt/issue_72181.rs" "/checkout/tests/mir-opt/retag.rs" "/checkout/tests/mir-opt/lower_intrinsics.rs" "/checkout/tests/mir-opt/issue_72181_1.rs" "/checkout/tests/mir-opt/unreachable_enum_branching.rs" "/checkout/tests/mir-opt/no_drop_for_inactive_variant.rs" "/checkout/tests/mir-opt/issue_120925_unsafefncast.rs" "/checkout/tests/mir-opt/lower_slice_len.rs" "/checkout/tests/mir-opt/remove_unneeded_drops.rs" "/checkout/tests/coverage-run-rustdoc/auxiliary/doctest_crate.rs" "/checkout/tests/coverage-run-rustdoc/doctest.rs" "/checkout/tests/rustdoc-js/search-short-types.rs" "/checkout/tests/rustdoc-js/never-search.rs" "/checkout/tests/rustdoc-js/doc-alias.rs" "/checkout/tests/rustdoc-js/trait-methods.rs" "/checkout/tests/rustdoc-js/non-english-identifier.rs" "/checkout/tests/rustdoc-js/foreign-type-path.rs" "/checkout/tests/rustdoc-js/doc-alias-whitespace.rs" "/checkout/tests/rustdoc-js/raw-pointer.rs" "/checkout/tests/rustdoc-js/assoc-type-backtrack.rs" "/checkout/tests/rustdoc-js/summaries.rs" "/checkout/tests/rustdoc-js/enum-variant-not-type.rs" "/checkout/tests/rustdoc-js/generics-trait.rs" "/checkout/tests/rustdoc-js/underscoredtype.rs" "/checkout/tests/rustdoc-js/reexport-dedup.rs" "/checkout/tests/rustdoc-js/self-is-not-generic.rs" "/checkout/tests/rustdoc-js/tuple-unit.rs" "/checkout/tests/rustdoc-js/generics.rs" "/checkout/tests/rustdoc-js/search-non-local-trait-impl.rs" "/checkout/tests/rustdoc-js/reexport.rs" "/checkout/tests/rustdoc-js/assoc-type-loop.rs" "/checkout/tests/rustdoc-js/auxiliary/macro-in-module.rs" "/checkout/tests/rustdoc-js/auxiliary/interner.rs" "/checkout/tests/rustdoc-js/auxiliary/equivalent.rs" "/checkout/tests/rustdoc-js/looks-like-rustc-interner.rs" "/checkout/tests/rustdoc-js/search-bag-semantics.rs" "/checkout/tests/rustdoc-js/struct-like-variant.rs" "/checkout/tests/rustdoc-js/nested-unboxed.rs" "/checkout/tests/rustdoc-js/generics-nested.rs" "/checkout/tests/rustdoc-js/primitive.rs" "/checkout/tests/rustdoc-js/doc-alias-filter-out.rs" "/checkout/tests/rustdoc-js/generics-multi-trait.rs" "/checkout/tests/rustdoc-js/generics-impl.rs" "/checkout/tests/rustdoc-js/path-maxeditdistance.rs" "/checkout/tests/rustdoc-js/type-parameters.rs" "/checkout/tests/mir-opt/building/match/never_patterns.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Tue Sep  3 13:50:55 UTC 2024
  network time: Tue, 03 Sep 2024 13:50:55 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #128299) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

DianQK commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 622247a04b8286df0e72a5f5dfa34dd74807b254 with merge c3241126e94d7a53a1805ad854686e0bd061b5df...
<!-- homu: {"type":"TryBuildStarted","head_sha":"622247a04b8286df0e72a5f5dfa34dd74807b254","merge_sha":"c3241126e94d7a53a1805ad854686e0bd061b5df"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10865820648/job/30153909819)
Build commit: c3241126e94d7a53a1805ad854686e0bd061b5df (`c3241126e94d7a53a1805ad854686e0bd061b5df`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10865820648/job/30153909819"},"merge_sha":"c3241126e94d7a53a1805ad854686e0bd061b5df"} -->

rust-timer commented:
Queued c3241126e94d7a53a1805ad854686e0bd061b5df with parent 5e3ede22ef6fd23197629c22b9f466b3734e19f1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5e3ede22ef6fd23197629c22b9f466b3734e19f1&end=c3241126e94d7a53a1805ad854686e0bd061b5df).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (c3241126e94d7a53a1805ad854686e0bd061b5df): [comparison URL](https://perf.rust-lang.org/compare.html?start=5e3ede22ef6fd23197629c22b9f466b3734e19f1&end=c3241126e94d7a53a1805ad854686e0bd061b5df&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=5e3ede22ef6fd23197629c22b9f466b3734e19f1&end=c3241126e94d7a53a1805ad854686e0bd061b5df&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5e3ede22ef6fd23197629c22b9f466b3734e19f1&end=c3241126e94d7a53a1805ad854686e0bd061b5df&stat=max-rss)
<details>
<summary>Results (primary 0.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.0%  | [3.0%, 5.1%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.8% | [-5.9%, -1.7%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.9%  | [-5.9%, 5.1%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=5e3ede22ef6fd23197629c22b9f466b3734e19f1&end=c3241126e94d7a53a1805ad854686e0bd061b5df&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=5e3ede22ef6fd23197629c22b9f466b3734e19f1&end=c3241126e94d7a53a1805ad854686e0bd061b5df&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary -0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.2%]   | 11    |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.1%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.3%, -0.0%] | 6     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 3     |
| All ❌✅ (primary)                 | 0.0%  | [-0.3%, 0.2%]  | 17    |
</details>

**Bootstrap**: 760.71s -> 756.595s (-0.54%)
**Artifact size**: 341.13 MiB -> 341.17 MiB (0.01%)

cjgillot commented on compiler/rustc_mir_transform/src/lib.rs:
```diff
@@ -604,6 +602,7 @@ fn run_optimization_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
             &dead_store_elimination::DeadStoreElimination::Initial,
             &gvn::GVN,
             &simplify::SimplifyLocals::AfterGVN,
+            &match_branches::MatchBranchSimplification,
```
Should this run on clone shims too? Or do we want a trait-based solution to detect trivial clone impls?

cjgillot commented on compiler/rustc_mir_transform/src/match_branches.rs:
```diff
@@ -48,6 +54,10 @@ impl<'tcx> crate::MirPass<'tcx> for MatchBranchSimplification {
+            if simplify_to_copy(tcx, body, bb_idx, param_env).is_some() {
+                should_cleanup = true;
+                continue;
+            }
```
Should this be made a standalone MirPass? This will be easier to have some analyses computed only once.

cjgillot commented on compiler/rustc_mir_transform/src/match_branches.rs:
```diff
@@ -519,3 +529,212 @@ impl<'tcx> SimplifyMatch<'tcx> for SimplifyToExp {
+        if stmts.is_empty() {
+            return None;
+        }
+        if let [Statement { kind: StatementKind::Assign(box (place, rvalue)), .. }] =
```
Could you add a comment for the high-level pattern you are trying to match?

cjgillot commented on compiler/rustc_mir_transform/src/match_branches.rs:
```diff
@@ -519,3 +529,212 @@ impl<'tcx> SimplifyMatch<'tcx> for SimplifyToExp {
+                    }
+                }
+            }
+            let expected_copy_stmt = expected_copy_stmt?;
```
Should we perform the same analysis on this statement that we do in the single statement branch above?

cjgillot reviewed (commented):
My high-level reaction to this PR is that it's both too specific in what it's trying to match (a clone impl) and too general in its implementation (handles storage statements...).

I suggest having two heuristics:
- for clone impls, use some kind of `StructuralClone` trait so that we fully replace the derived impls `Clone` with simpler MIR;
- for general MIR, lift the restrictions on this pass (bb0 and assignment to _0 namely).

What do you think?

DianQK commented:
>     * for clone impls, use some kind of `StructuralClone` trait so that we fully replace the derived impls `Clone` with simpler MIR;
> 
>     * for general MIR, lift the restrictions on this pass (bb0 and assignment to _0 namely).
> 
> What do you think?

It looks like makes sense.
IIUC, the `StructuralClone` is exact matching clone shims? I expect this can improve compile-time.

@rustbot author

dianqk reviewed (commented):
I didn't submit all the comments through "Submit review"; there are still some pending comments. `o`

I am currently modifying the code to move it into a separate pass.

DianQK commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit dddf04628e1e3e9ecc51b06d3791c65ebbc4ba3f with merge 366e98f4a5ed3bd1a9882e096b12eba79e126ddc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dddf04628e1e3e9ecc51b06d3791c65ebbc4ba3f","merge_sha":"366e98f4a5ed3bd1a9882e096b12eba79e126ddc"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11193600960/job/31120432973)
Build commit: 366e98f4a5ed3bd1a9882e096b12eba79e126ddc (`366e98f4a5ed3bd1a9882e096b12eba79e126ddc`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11193600960/job/31120432973"},"merge_sha":"366e98f4a5ed3bd1a9882e096b12eba79e126ddc"} -->

rust-timer commented:
Queued 366e98f4a5ed3bd1a9882e096b12eba79e126ddc with parent 5a4ee43c387110736440cecc375cb5aedc441dbc, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5a4ee43c387110736440cecc375cb5aedc441dbc&end=366e98f4a5ed3bd1a9882e096b12eba79e126ddc).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (366e98f4a5ed3bd1a9882e096b12eba79e126ddc): [comparison URL](https://perf.rust-lang.org/compare.html?start=5a4ee43c387110736440cecc375cb5aedc441dbc&end=366e98f4a5ed3bd1a9882e096b12eba79e126ddc&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=5a4ee43c387110736440cecc375cb5aedc441dbc&end=366e98f4a5ed3bd1a9882e096b12eba79e126ddc&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4% | [0.2%, 0.8%] | 6     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.4% | [0.2%, 0.8%] | 6     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5a4ee43c387110736440cecc375cb5aedc441dbc&end=366e98f4a5ed3bd1a9882e096b12eba79e126ddc&stat=max-rss)
<details>
<summary>Results (primary 2.8%, secondary -3.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.3%, 3.4%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.1% | [-3.1%, -3.1%] | 1     |
| All ❌✅ (primary)                 | 2.8%  | [2.3%, 3.4%]   | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=5a4ee43c387110736440cecc375cb5aedc441dbc&end=366e98f4a5ed3bd1a9882e096b12eba79e126ddc&stat=cycles:u)
<details>
<summary>Results (secondary -1.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [2.0%, 2.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.6%, -2.3%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=5a4ee43c387110736440cecc375cb5aedc441dbc&end=366e98f4a5ed3bd1a9882e096b12eba79e126ddc&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary -0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.1%, 0.4%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.4%, -0.0%] | 8     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | -0.0% | [-0.4%, 0.4%]  | 11    |
</details>

**Bootstrap**: 774.178s -> 775.458s (0.17%)
**Artifact size**: 342.23 MiB -> 342.33 MiB (0.03%)

DianQK commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 7846743c0a2634d6eba9889a5b411307f792fbde with merge 963788449d01d434fa9a69b331e4f5c41f10eab4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7846743c0a2634d6eba9889a5b411307f792fbde","merge_sha":"963788449d01d434fa9a69b331e4f5c41f10eab4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11201700127/job/31138543389)
Build commit: 963788449d01d434fa9a69b331e4f5c41f10eab4 (`963788449d01d434fa9a69b331e4f5c41f10eab4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11201700127/job/31138543389"},"merge_sha":"963788449d01d434fa9a69b331e4f5c41f10eab4"} -->

rust-timer commented:
Finished benchmarking commit (963788449d01d434fa9a69b331e4f5c41f10eab4): [comparison URL](https://perf.rust-lang.org/compare.html?start=68301a6a968326bac8b85b04d5ef7e79cc042a00&end=963788449d01d434fa9a69b331e4f5c41f10eab4&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=68301a6a968326bac8b85b04d5ef7e79cc042a00&end=963788449d01d434fa9a69b331e4f5c41f10eab4&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.3%, 0.8%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 0.5%  | [0.5%, 0.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.5% | [-0.5%, -0.5%] | 1     |
| All ❌✅ (primary)                 | 0.5%  | [0.3%, 0.8%]   | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=68301a6a968326bac8b85b04d5ef7e79cc042a00&end=963788449d01d434fa9a69b331e4f5c41f10eab4&stat=max-rss)
<details>
<summary>Results (primary 3.2%, secondary 2.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.2% | [2.2%, 4.7%] | 4     |
| Regressions ❌ <br /> (secondary)  | 2.5% | [2.5%, 2.5%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.2% | [2.2%, 4.7%] | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=68301a6a968326bac8b85b04d5ef7e79cc042a00&end=963788449d01d434fa9a69b331e4f5c41f10eab4&stat=cycles:u)
<details>
<summary>Results (primary 1.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.9% | [1.0%, 2.8%] | 2     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.9% | [1.0%, 2.8%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=68301a6a968326bac8b85b04d5ef7e79cc042a00&end=963788449d01d434fa9a69b331e4f5c41f10eab4&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary -0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.1%, 0.4%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.4%, -0.0%] | 8     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | -0.0% | [-0.4%, 0.4%]  | 11    |
</details>

**Bootstrap**: 775.25s -> 775.189s (-0.01%)
**Artifact size**: 329.52 MiB -> 329.45 MiB (-0.02%)

DianQK commented:
@rustbot review

cjgillot commented on compiler/rustc_mir_transform/src/merge_branches.rs:
```diff
@@ -0,0 +1,289 @@
+        }
+        expected_assign
+    };
+    let Statement { kind: StatementKind::Assign(box (dest_place, ref rvalue)), .. } =
+        statements[assign_stmt]
+    else {
+        return None;
+    };
```
```suggestion
    let &(dest_place, ref rvalue) = statements[assign_stmt].kind.as_assign()?;
```

cjgillot commented on compiler/rustc_mir_transform/src/lib.rs:
```diff
@@ -609,6 +610,7 @@ fn run_optimization_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
             &dead_store_elimination::DeadStoreElimination::Initial,
             &gvn::GVN,
             &simplify::SimplifyLocals::AfterGVN,
+            &merge_branches::MergeBranchSimplification,
```
Why here in the pipeline?

DianQK replied:
This needs to run after GVN. I expect `SimplifyLocals::AfterGVN` can simplify the analysis in `MergeBranchSimplification`. Additionally, after `MergeBranchSimplification` simplifies the CFG, it will benefit subsequent passes that are related to the CFG.

cjgillot replied:
Should we add a `dead_store_elimination::DeadStoreElimination::AfterGVN` just before it and simplify your pass?

dianqk replied:
I don't want DSE to modify CFG. I have made code adjustments and merged them into `MatchBranchSimplification`.

DianQK commented on compiler/rustc_mir_transform/src/merge_branches.rs:
```diff
@@ -0,0 +1,289 @@
+                        // There is only one statement that cannot be ignored
+                        // that can be used as an expected copy statement.
+                        expected_assign_stmt = Some(statement_index);
+                        lived_stmts.remove(statement_index);
```
Since debuginfo handles it differently, I'll try to modify the DSE then get rid of these.

DianQK commented:
I will probably update in two days.

@rustbot author

DianQK commented:
Just [a few trivial modifications](https://github.com/rust-lang/rust/compare/e1d9550512407a60c718eb91874c588fca908caa..c25e375cd051aa3696518224578d7d8a5968a878) and rebase.

DianQK commented on tests/mir-opt/merge_br/merge_br.rs:
```diff
@@ -0,0 +1,90 @@
+    // CHECK: bb0: {
+    // CHECK-NEXT: _{{.*}} = discriminant([[SRC:_1]]);
+    // CHECK-NEXT: _0 = copy [[SRC]];
+    // CHECK-NEXT return;
```
I replaced `simplify_cfg` with `remove_dead_blocks`, so I just deleted these.

DianQK commented on compiler/rustc_mir_transform/src/merge_branches.rs:
```diff
@@ -0,0 +1,289 @@
+                continue;
+            }
+            let src_ty = src_place.ty(body.local_decls(), tcx);
+            if !src_ty.ty.is_enum() || src_ty.variant_index.is_some() {
```
Ah, yes!

DianQK commented:
I have tried to update the document and DSE.

@rustbot review

bors commented:
:umbrella: The latest upstream changes (presumably #132349) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

DianQK commented:
Blocked by #132353.


@rustbot author
@rustbot label +S-blocked

bors commented:
:hourglass: Trying commit f05d4af6be342335fdf4f4c1f765fe31d780e73f with merge 27bc686b0d1e864190a5a718151fd4ddcea59971...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f05d4af6be342335fdf4f4c1f765fe31d780e73f","merge_sha":"27bc686b0d1e864190a5a718151fd4ddcea59971"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14263008852/job/39985368693)
Build commit: 27bc686b0d1e864190a5a718151fd4ddcea59971 (`27bc686b0d1e864190a5a718151fd4ddcea59971`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14263008852/job/39985368693"},"merge_sha":"27bc686b0d1e864190a5a718151fd4ddcea59971"} -->

rust-timer commented:
Queued 27bc686b0d1e864190a5a718151fd4ddcea59971 with parent f174fd716a429fa17eb1e98ba4e382f09312f8ad, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f174fd716a429fa17eb1e98ba4e382f09312f8ad&end=27bc686b0d1e864190a5a718151fd4ddcea59971).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (27bc686b0d1e864190a5a718151fd4ddcea59971): [comparison URL](https://perf.rust-lang.org/compare.html?start=f174fd716a429fa17eb1e98ba4e382f09312f8ad&end=27bc686b0d1e864190a5a718151fd4ddcea59971&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=f174fd716a429fa17eb1e98ba4e382f09312f8ad&end=27bc686b0d1e864190a5a718151fd4ddcea59971&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.5%]   | 6     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.4% | [-1.4%, -1.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-1.4%, 0.5%]  | 7     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f174fd716a429fa17eb1e98ba4e382f09312f8ad&end=27bc686b0d1e864190a5a718151fd4ddcea59971&stat=max-rss)
<details>
<summary>Results (primary -3.8%, secondary -2.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.8% | [-5.6%, -2.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | -3.8% | [-5.6%, -2.1%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f174fd716a429fa17eb1e98ba4e382f09312f8ad&end=27bc686b0d1e864190a5a718151fd4ddcea59971&stat=cycles:u)
<details>
<summary>Results (primary 1.8%, secondary 2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.8% | [1.8%, 1.8%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.3% | [2.3%, 2.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.8% | [1.8%, 1.8%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f174fd716a429fa17eb1e98ba4e382f09312f8ad&end=27bc686b0d1e864190a5a718151fd4ddcea59971&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.1%, secondary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.1%, 0.6%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.9%, -0.0%] | 17    |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.1%] | 35    |
| All ❌✅ (primary)                 | -0.1% | [-0.9%, 0.6%]  | 20    |
</details>

**Bootstrap**: 778.988s -> 776.478s (-0.32%)
**Artifact size**: 365.99 MiB -> 366.08 MiB (0.02%)

cjgillot commented on compiler/rustc_mir_transform/src/merge_branches.rs:
```diff
@@ -0,0 +1,243 @@
+            ) {
+                let statement_index = bbs[switch_bb_idx].statements.len();
+                let parent_end = Location { block: switch_bb_idx, statement_index };
+                let mut patch = MirPatch::new(body);
+                patch.add_assign(parent_end, dest_place, Rvalue::Use(Operand::Copy(src_place)));
+                patch.patch_terminator(switch_bb_idx, first_terminator_kind.clone());
+                patch.apply(body);
+                super::simplify::remove_dead_blocks(body);
+                // After modifying the MIR, the result of `MaybeTransitiveLiveLocals` may become invalid,
+                // keeping it simple to process only once.
+                break;
```
I'd recommend having a single patch and modify MIR at the end. This allows to catch several instances of the opt in the same body.

dianqk replied:
I fixed this in [8105aefe072e1dffbcc2f0d9114e48c9add6d03b](https://github.com/rust-lang/rust/pull/129931/commits/8105aefe072e1dffbcc2f0d9114e48c9add6d03b).

cjgillot commented on compiler/rustc_mir_transform/src/merge_branches.rs:
```diff
@@ -0,0 +1,243 @@
+        let mut lived_stmts: DenseBitSet<usize> = DenseBitSet::new_filled(statements.len());
+        let mut expected_assign_stmt = None;
+        for (statement_index, statement) in statements.iter().enumerate().rev() {
+            let loc = Location { block: target_block, statement_index };
+            if dead_store_analysis.is_dead_store(loc, &statement.kind) {
+                lived_stmts.remove(statement_index);
+            } else if matches!(
+                statement.kind,
+                StatementKind::StorageLive(_) | StatementKind::StorageDead(_)
+            ) {
+            } else if matches!(statement.kind, StatementKind::Assign(_))
+                && expected_assign_stmt.is_none()
+            {
+                // There is only one assign statement that cannot be ignored
+                // that can be used as an expected copy statement.
+                expected_assign_stmt = Some(statement_index);
+                lived_stmts.remove(statement_index);
+            } else {
+                return None;
+            }
+        }
+        let expected_assign = expected_assign_stmt?;
+        if !lived_stmts.is_empty() {
+            // We can ignore the paired StorageLive and StorageDead.
+            let mut storage_live_locals: DenseBitSet<Local> =
+                DenseBitSet::new_empty(body.local_decls.len());
+            for stmt_index in lived_stmts.iter() {
+                let statement = &statements[stmt_index];
+                match &statement.kind {
+                    StatementKind::StorageLive(local) if storage_live_locals.insert(*local) => {}
+                    StatementKind::StorageDead(local) if storage_live_locals.remove(*local) => {}
+                    _ => return None,
+                }
+            }
+            if !storage_live_locals.is_empty() {
+                return None;
+            }
+        }
+        expected_assign
+    };
```
Could you make this a standalone method?

dianqk replied:
I'm no longer using these codes, the new patch can handle them better.

cjgillot commented on compiler/rustc_mir_transform/src/merge_branches.rs:
```diff
@@ -0,0 +1,243 @@
+    body: &Body<'tcx>,
+    targets: &SwitchTargets,
+    src_place: Place<'tcx>,
+    src_ty: PlaceTy<'tcx>,
```
This can be computed from `src_place` inside the function.

cjgillot commented on compiler/rustc_mir_transform/src/merge_branches.rs:
```diff
@@ -0,0 +1,243 @@
+                return None;
+            }
+        }
+        Rvalue::Use(Operand::Copy(place)) if *place == src_place => {}
```
What about `Operand::Move`? Downgrading a move to a copy is fine.

dianqk replied:
I'll address it in the next PR or a subsequent update.

dianqk committed:
mir-opt: Use one MirPatch in MatchBranchSimplification

dianqk committed:
mir-opt: ignore dead store stmts in MatchBranchSimplification

dianqk committed:
mir-opt: run MatchBranchSimplification after GVN

dianqk commented:
To better handle dead store statements, I'm cloning basic blocks in `MatchBranchSimplification` and performing DSE. I think this approach is more generic, and easier to maintain and review.

I haven't transformed `match a { Foo::A(_x) => a, Foo::B => Foo::B }` to a copy statement yet this weekend, but I think this can be a subsequent PR.

@rustbot review

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14292239493/job/40054944148) [(plain)](https://github.com/rust-lang/rust/commit/543cf865ad9b0b6b0c792c7d16eb86ba389273a0/checks/40054944148/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Mismatch at tests/source/struct-field-attributes.rs:24:
 // #1462
 struct Foo {
     foo: usize,
-    #[cfg(feature = "include-bar")]
+    #[cfg(feature="include-bar")]
     bar: usize,
 }
 

Mismatch at tests/source/struct-field-attributes.rs:39:
 // #2044
 pub enum State {
     Closure(
-        #[cfg_attr(
-            feature = "serde_derive",
-            serde(state_with = "::serialization::closure")
-        )]
-        GcPtr<ClosureData>,
+        #[cfg_attr(feature = "serde_derive", serde(state_with = "::serialization::closure"))] GcPtr<ClosureData>,
     ),
 }
 

Mismatch at tests/source/struct-field-attributes.rs:50:
 struct Fields(
-    #[cfg_attr(
-        feature = "serde_derive",
-        serde(state_with = "::base::serialization::shared")
-    )]
-    Arc<Vec<InternedStr>>,
+    #[cfg_attr(feature = "serde_derive", serde(state_with = "::base::serialization::shared"))] Arc<Vec<InternedStr>>,
 );
 
 // #2309

Mismatch at tests/source/struct-field-attributes.rs:59:
 pub struct A {
-    #[doc = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"]
-    pub foos: Vec<bool>,
+    #[doc="XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"]
+pub foos:Vec<bool>,
 }
 

Mismatch at tests/source/unions.rs:32:
 // #410
 #[allow(missing_docs)]
 pub union Writebatch<K: Key> {
-    #[allow(dead_code)] // only used for holding the internal pointer
+    #[allow(dead_code)] //only used for holding the internal pointer
     writebatch: RawWritebatch,
     marker: PhantomData<K>,
 }

Mismatch at tests/source/unions.rs:91:
 }
 
 union Palette {
-    /// A map of indices in the palette to a count of pixels in approximately
-    /// that color
-    foo: i32,
+    /// A map of indices in the palette to a count of pixels in approximately that color
+                    foo: i32,
 }
 
 // Splitting a single line comment into a block previously had a misalignment

Mismatch at tests/source/unions.rs:104:
                                        * BBBBBBBBBBBBBBBB BBBBBBBBBBBBBBBBB BBBBBBBBBBB */
     // Another pre comment
     #[attr1]
-    #[attr2]
-    pub id: usize, /* CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCC
-                    * CCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCC CCCCCCCCCCCC */
+    #[attr2] pub id: usize, /* CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCC
+                             * CCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCC CCCCCCCCCCCC */
 }
 
 union Deep {

Mismatch at tests/source/unions.rs:113:
-    deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep:
-        node::Handle<IdRef<'id, Node<K, V>>, Type, NodeType>,
+    deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep: node::Handle<IdRef<'id, Node<K, V>>,
+                                                     Type,
+                                                     NodeType>,
 }
 
 mod m {

Mismatch at tests/source/unions.rs:125:
 
 union Issue677 {
     pub ptr: *const libc::c_void,
-    pub trace: fn(obj: *const libc::c_void, tracer: *mut JSTracer),
+    pub trace: fn(  obj: 
+          *const libc::c_void, tracer   : *mut   JSTracer   ),
 }
 
 union Foo {}

Mismatch at tests/source/unions.rs:142:
 
 union LongUnion {
     a: A,
-    the_quick_brown_fox_jumps_over_the_lazy_dog:
-        AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA,
+    the_quick_brown_fox_jumps_over_the_lazy_dog:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA,
 }
 
 union Deep {

Mismatch at tests/source/unions.rs:150:
-    deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep:
-        node::Handle<IdRef<'id, Node<Key, Value>>, Type, NodeType>,
+    deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep: node::Handle<IdRef<'id, Node<Key, Value>>,
+                                                                         Type,
+                                                                         NodeType>,
 }
 
 // #1364

Mismatch at tests/source/issue_5686.rs:34:
     #[hell{world}]
     Cobra {/* struct variant close in leading attribute */} = 6,
     Eagle {
-        /* struct variant closer on associated field attribute */
-        #[hell{world}]
-        value: Sting,
+        /* struct variant closer on associated field attribute */ #[hell{world}]value: Sting,
     } = 7,
     Koala {/* struct variant closer on macro call */} = some_macro! {},
 }

Mismatch at tests/source/multiple.rs:105:
 pub struct Foo {
     #[rustfmt::skip]
     f :   SomeType, // Comment beside a field
-    f: SomeType, // Comment beside a field
+    f :   SomeType, // Comment beside a field
     // Comment on a field
     g: SomeOtherType,
     /// A doc comment on a field

Mismatch at tests/source/issue-2869.rs:3:
 #[derive(Serialize, Deserialize, Debug)]
 #[serde(rename_all = "PascalCase")]
---
-    result_status:                     Option<String>,
+    creation_time: String,
+    id: String,
+    operation: String,
+    organization_id: String,
+    record_type: u32,
+    result_status: Option<String>,
     #[serde(rename = "ClientIP")]
-    client_ip:                         Option<IpAddr>,
-    object_id:                         String,
-    actor:                             Option<Vec<IDType>>,
-    actor_context_id:                  Option<String>,
-    actor_ip_address:                  Option<IpAddr>,
+    client_ip: Option<IpAddr>,
+    object_id: String,
+    actor: Option<Vec<IDType>>,
+    actor_context_id: Option<String>,
+    actor_ip_address: Option<IpAddr>,
     azure_active_directory_event_type: Option<u8>,
 
     #[serde(rename = "very")]

Mismatch at tests/source/issue-2869.rs:21:
     aaaaa: String,
---
 struct AuditLog2 {
-    creation_time:                     String,
-    id:                                String,
-    operation:                         String,
-    organization_id:                   String,
-    record_type:                       u32,
-    result_status:                     Option<String>,
-    client_ip:                         Option<IpAddr>,
-    object_id:                         String,
-    actor:                             Option<Vec<IDType>>,
-    actor_context_id:                  Option<String>,
-    actor_ip_address:                  Option<IpAddr>,
+    creation_time: String,
+    id: String,
+    operation: String,
+    organization_id: String,
+    record_type: u32,
+    result_status: Option<String>,
+    client_ip: Option<IpAddr>,
+    object_id: String,
+    actor: Option<Vec<IDType>>,
+    actor_context_id: Option<String>,
+    actor_ip_address: Option<IpAddr>,
     azure_active_directory_event_type: Option<u8>,
 }
 

Mismatch at tests/source/issue-2164.rs:2:
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
 pub struct emacs_env_25 {
-    pub size: isize,
-    pub private_members: *mut emacs_env_private,
-    pub make_global_ref: ::std::option::Option<
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14292625375/job/40055833653) [(plain)](https://github.com/rust-lang/rust/commit/c91a07a3f4da0e5c07b5bca0ce6ef532496ff974/checks/40055833653/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Mismatch at tests/source/struct-field-attributes.rs:24:
 // #1462
 struct Foo {
     foo: usize,
-    #[cfg(feature = "include-bar")]
+    #[cfg(feature="include-bar")]
     bar: usize,
 }
 

Mismatch at tests/source/struct-field-attributes.rs:39:
 // #2044
 pub enum State {
     Closure(
-        #[cfg_attr(
-            feature = "serde_derive",
-            serde(state_with = "::serialization::closure")
-        )]
-        GcPtr<ClosureData>,
+        #[cfg_attr(feature = "serde_derive", serde(state_with = "::serialization::closure"))] GcPtr<ClosureData>,
     ),
 }
 

Mismatch at tests/source/struct-field-attributes.rs:50:
 struct Fields(
-    #[cfg_attr(
-        feature = "serde_derive",
-        serde(state_with = "::base::serialization::shared")
-    )]
-    Arc<Vec<InternedStr>>,
+    #[cfg_attr(feature = "serde_derive", serde(state_with = "::base::serialization::shared"))] Arc<Vec<InternedStr>>,
 );
 
 // #2309

Mismatch at tests/source/struct-field-attributes.rs:59:
 pub struct A {
-    #[doc = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"]
-    pub foos: Vec<bool>,
+    #[doc="XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"]
+pub foos:Vec<bool>,
 }
 

Mismatch at tests/source/unions.rs:32:
 // #410
 #[allow(missing_docs)]
 pub union Writebatch<K: Key> {
-    #[allow(dead_code)] // only used for holding the internal pointer
+    #[allow(dead_code)] //only used for holding the internal pointer
     writebatch: RawWritebatch,
     marker: PhantomData<K>,
 }

Mismatch at tests/source/unions.rs:91:
 }
 
 union Palette {
-    /// A map of indices in the palette to a count of pixels in approximately
-    /// that color
-    foo: i32,
+    /// A map of indices in the palette to a count of pixels in approximately that color
+                    foo: i32,
 }
 
 // Splitting a single line comment into a block previously had a misalignment

Mismatch at tests/source/unions.rs:104:
                                        * BBBBBBBBBBBBBBBB BBBBBBBBBBBBBBBBB BBBBBBBBBBB */
     // Another pre comment
     #[attr1]
-    #[attr2]
-    pub id: usize, /* CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCC
-                    * CCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCC CCCCCCCCCCCC */
+    #[attr2] pub id: usize, /* CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCC
+                             * CCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCC CCCCCCCCCCCC */
 }
 
 union Deep {

Mismatch at tests/source/unions.rs:113:
-    deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep:
-        node::Handle<IdRef<'id, Node<K, V>>, Type, NodeType>,
+    deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep: node::Handle<IdRef<'id, Node<K, V>>,
+                                                     Type,
+                                                     NodeType>,
 }
 
 mod m {

Mismatch at tests/source/unions.rs:125:
 
 union Issue677 {
     pub ptr: *const libc::c_void,
-    pub trace: fn(obj: *const libc::c_void, tracer: *mut JSTracer),
+    pub trace: fn(  obj: 
+          *const libc::c_void, tracer   : *mut   JSTracer   ),
 }
 
 union Foo {}

Mismatch at tests/source/unions.rs:142:
 
 union LongUnion {
     a: A,
-    the_quick_brown_fox_jumps_over_the_lazy_dog:
-        AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA,
+    the_quick_brown_fox_jumps_over_the_lazy_dog:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA,
 }
 
 union Deep {

Mismatch at tests/source/unions.rs:150:
-    deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep:
-        node::Handle<IdRef<'id, Node<Key, Value>>, Type, NodeType>,
+    deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep: node::Handle<IdRef<'id, Node<Key, Value>>,
+                                                                         Type,
+                                                                         NodeType>,
 }
 
 // #1364

Mismatch at tests/source/issue_5686.rs:34:
     #[hell{world}]
     Cobra {/* struct variant close in leading attribute */} = 6,
     Eagle {
-        /* struct variant closer on associated field attribute */
-        #[hell{world}]
-        value: Sting,
+        /* struct variant closer on associated field attribute */ #[hell{world}]value: Sting,
     } = 7,
     Koala {/* struct variant closer on macro call */} = some_macro! {},
 }

Mismatch at tests/source/multiple.rs:105:
 pub struct Foo {
     #[rustfmt::skip]
     f :   SomeType, // Comment beside a field
-    f: SomeType, // Comment beside a field
+    f :   SomeType, // Comment beside a field
     // Comment on a field
     g: SomeOtherType,
     /// A doc comment on a field

Mismatch at tests/source/issue-2869.rs:3:
 #[derive(Serialize, Deserialize, Debug)]
 #[serde(rename_all = "PascalCase")]
---
-    result_status:                     Option<String>,
+    creation_time: String,
+    id: String,
+    operation: String,
+    organization_id: String,
+    record_type: u32,
+    result_status: Option<String>,
     #[serde(rename = "ClientIP")]
-    client_ip:                         Option<IpAddr>,
-    object_id:                         String,
-    actor:                             Option<Vec<IDType>>,
-    actor_context_id:                  Option<String>,
-    actor_ip_address:                  Option<IpAddr>,
+    client_ip: Option<IpAddr>,
+    object_id: String,
+    actor: Option<Vec<IDType>>,
+    actor_context_id: Option<String>,
+    actor_ip_address: Option<IpAddr>,
     azure_active_directory_event_type: Option<u8>,
 
     #[serde(rename = "very")]

Mismatch at tests/source/issue-2869.rs:21:
     aaaaa: String,
---
 struct AuditLog2 {
-    creation_time:                     String,
-    id:                                String,
-    operation:                         String,
-    organization_id:                   String,
-    record_type:                       u32,
-    result_status:                     Option<String>,
-    client_ip:                         Option<IpAddr>,
-    object_id:                         String,
-    actor:                             Option<Vec<IDType>>,
-    actor_context_id:                  Option<String>,
-    actor_ip_address:                  Option<IpAddr>,
+    creation_time: String,
+    id: String,
+    operation: String,
+    organization_id: String,
+    record_type: u32,
+    result_status: Option<String>,
+    client_ip: Option<IpAddr>,
+    object_id: String,
+    actor: Option<Vec<IDType>>,
+    actor_context_id: Option<String>,
+    actor_ip_address: Option<IpAddr>,
     azure_active_directory_event_type: Option<u8>,
 }
 

Mismatch at tests/source/issue-2164.rs:2:
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
 pub struct emacs_env_25 {
-    pub size: isize,
-    pub private_members: *mut emacs_env_private,
-    pub make_global_ref: ::std::option::Option<
```

</details>

dianqk commented:
> The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14292625375/job/40055833653) [(plain)](https://github.com/rust-lang/rust/commit/c91a07a3f4da0e5c07b5bca0ce6ef532496ff974/checks/40055833653/logs)
> 
> Click to see the possible cause of the failure (guessed by this bot)

@rustbot author

dianqk committed:
mir-opt: Do not eliminate dead statements that are being used by other BBs during merging

Since we will not eliminate dead statements in other BBs,
they may still use the dead statements within the merged BB.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14368298554/job/40286272983) [(plain)](https://github.com/rust-lang/rust/commit/768b800605cdacd5dab08b268770c19c3dbe3aa1/checks/40286272983/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test test::format_lines_errors_are_reported_with_tabs ... ok
error: test failed, to rerun pass `--lib`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/rustfmt_nightly-4ff62fb8390505dc -Z unstable-options --format json` (signal: 11, SIGSEGV: invalid memory reference)
Build completed unsuccessfully in 0:01:14
  local time: Wed Apr  9 23:07:25 UTC 2025
  network time: Wed, 09 Apr 2025 23:07:26 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

dianqk commented:
> Should we add a `dead_store_elimination::DeadStoreElimination::AfterGVN` just before it and simplify your pass?

I will add it and let it work with debuginfo. It seems I need to address #68817 first.
